### PR TITLE
[Semi Auto] Softmax SPMD Rule

### PR DESCRIPTION
**Pull request by JZ-LIANG**
### PR types
Function optimization

### PR changes
Others

### Description
Pcard-70448

SPMD rule for softmax like ops (softmax, log_softmax).
sharding on normlized axis will be support in future.

### CODE REVIEW

File: paddle/fluid/distributed/auto_parallel/spmd_rules/embedding_spmd_rule.cc [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+      x_dims_mapping.size(),
+      phi::errors::InvalidArgument(
+          "Mismatch of X's tensor size: [%d] and X's dims_mapping size [%d].",
+          x_ndim,
+          x_dims_mapping.size()));
+  PADDLE_ENFORCE_EQ(
+      weight_ndim,
+      weight_dims_mapping.size(),
+      phi::errors::InvalidArgument(
+          "Mismatch of Y's tensor size: [%d] and Y's dims_mapping size [%d].",
```
Comment (zhiqiu) [inline paddle/fluid/distributed/auto_parallel/spmd_rules/embedding_spmd_rule.cc]: Y -> w?

File: paddle/fluid/distributed/auto_parallel/spmd_rules/embedding_spmd_rule.cc [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+                                   weight_ndim));
+
+  int64_t padding_idx = ExtractAttr<int64_t>("padding_idx", attrs);
+  bool sparse = ExtractAttr<bool>("sparse", attrs);
+
+  // determine parallel mode
+  int64_t weight_row_axis_mapping = weight_dims_mapping[0];
+
+  // padding_idx s not supported by c_embedding kernl.
+  // (TODO) might be could reshard as replicated when padding_idx != -1
```
Comment (zhiqiu) [inline paddle/fluid/distributed/auto_parallel/spmd_rules/embedding_spmd_rule.cc]: hard to understand

File: paddle/fluid/distributed/auto_parallel/spmd_rules/embedding_spmd_rule.cc [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+          << "X shape: [" << str_join(x_shape) << "], src_dims_mapping: ["
+          << str_join(x_dims_mapping) << "], dst_dims_mapping: ["
+          << str_join(x_dims_mapping) << "]; Y shape: ["
+          << str_join(weight_shape) << "], src_dims_mapping: ["
+          << str_join(weight_dims_mapping) << "], dst_dims_mapping: ["
+          << str_join(weight_dims_mapping) << "]; Output dims_mapping: ["
+          << str_join(out_dims_mapping) << "], partial_on_dims: ["
+          << str_join(partial_on_dims) << "]";
+
+  return {{x_dist_attr_src, weight_dist_attr_src}, {output_dist_attr_dst}};
```
Comment (zhiqiu) [inline paddle/fluid/distributed/auto_parallel/spmd_rules/embedding_spmd_rule.cc]: infer inputs?

Comment (JZ-LIANG) [reply paddle/fluid/distributed/auto_parallel/spmd_rules/embedding_spmd_rule.cc]: done

File: paddle/fluid/distributed/auto_parallel/spmd_rules/embedding_spmd_rule.cc (hunk 1 above)
Comment (JZ-LIANG) [reply paddle/fluid/distributed/auto_parallel/spmd_rules/embedding_spmd_rule.cc]: fixed

File: paddle/fluid/distributed/auto_parallel/spmd_rules/embedding_spmd_rule.cc (hunk 2 above)
Comment (JZ-LIANG) [reply paddle/fluid/distributed/auto_parallel/spmd_rules/embedding_spmd_rule.cc]: the implement of  c_embedding kernel  require the padding item is in the of table, which means "padding_idx = -1" and not allow being set by user.

to support vocab parallel of embedding op efficiently, we need to use c_embedding kernel, therefore, we make a precondition check that "padding_idx = -1" when embedding table is sharded by vocab axis.

**Review by zhiqiu — APPROVED** [review]
LGTM

**Pull request merged**